alibi/explainers/__init__.py

```diff
@@ -2,34 +2,53 @@
 The 'alibi.explainers' module includes feature importance, counterfactual and anchor-based explainers.
 """
 
+from alibi.utils.missing_optional_dependency import import_optional
 from .ale import ALE, plot_ale
-from .anchor_tabular import AnchorTabular, DistributedAnchorTabular
-from .anchor_text import AnchorText
-from .anchor_image import AnchorImage
-from .cem import CEM
-from .cfproto import CounterfactualProto, CounterFactualProto  # noqa: F401 TODO: remove in an upcoming release
-from .counterfactual import Counterfactual, CounterFactual  # noqa: F401 TODO: remove in an upcoming release
-from .integrated_gradients import IntegratedGradients
-from .cfrl_base import CounterfactualRL
+from alibi.explainers.anchors.anchor_text import AnchorText
+from alibi.explainers.anchors.anchor_tabular import AnchorTabular
+from alibi.explainers.anchors.anchor_image import AnchorImage
+from .cfrl_base import CounterfactualRL, Callback
 from .cfrl_tabular import CounterfactualRLTabular
 
-__all__ = ["ALE",
-           "AnchorTabular",
-           "DistributedAnchorTabular",
-           "AnchorText",
-           "AnchorImage",
-           "CEM",
-           "Counterfactual",
-           "CounterfactualProto",
-           "CounterfactualRL",
-           "CounterfactualRLTabular",
-           "plot_ale",
-           "IntegratedGradients",
-           ]
-
-try:
-    from .shap_wrappers import KernelShap, TreeShap
-
-    __all__ += ["KernelShap", "TreeShap"]
-except ImportError:
-    pass
+
+DistributedAnchorTabular = import_optional(
+    'alibi.explainers.anchors.anchor_tabular_distributed',
+    names=['DistributedAnchorTabular'])
+
+CEM = import_optional(
+    'alibi.explainers.cem',
+    names=['CEM'])
+
+CounterfactualProto, CounterFactualProto = import_optional(
+    'alibi.explainers.cfproto',
+    names=['CounterfactualProto', 'CounterFactualProto'])  # TODO: remove in an upcoming release
+
+Counterfactual, CounterFactual = import_optional(
+    'alibi.explainers.counterfactual',
+    names=['Counterfactual', 'CounterFactual'])  # TODO: remove in an upcoming release
```

Comment:
What I noticed that doing `import CounterFactual` with `tensorflow` not installed doesn't raise the name deprecation warning. I don't think this is an issue though, just noting it.